Assets/MixedRealityToolkit.Services/InputSimulation/Implementation/MouseRotationProvider.cs

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.Input
+{
+    /// <summary>
+    /// Utility class to manage toggling of mouse rotation and associated features,
+    /// such as cursor visibility/locking
+    /// </summary>
+    /// <param name="wantsJumping">Show the cursor</param>
+    public class MouseRotationProvider
+    {
+        private bool isRotating = false;
+        public bool IsRotating => isRotating;
+
+        private static int numRotating = 0;
+        private static bool isMouseJumping = false;
+        private static bool wasCursorVisible = true;
+
+        public void Update(KeyBinding rotationKey, KeyBinding cancelRotationKey, bool toggle)
+        {
+            bool wasRotating = isRotating;
+
+            // Only allow the mouse to control rotation when Unity has focus.
+            // This enables the player to temporarily alt-tab away without having the player
+            // look around randomly back in the Unity Game window.
+            if (!Application.isFocused)
+            {
+                isRotating = false;
+            }
+            else
+            {
+                if (toggle)
+                {
+                    if (isRotating)
+                    {
+                        // pressing escape will stop capture
+                        isRotating = !KeyInputSystem.GetKeyDown(cancelRotationKey);
+                    }
+                    else
+                    {
+                        // any kind of click will capture focus
+                        isRotating = KeyInputSystem.GetKeyDown(rotationKey);
+                    }
+                }
+                else
+                {
+                    isRotating = KeyInputSystem.GetKey(rotationKey);
+                }
+            }
+
+            if (!wasRotating && isRotating)
+            {
+                OnStartRotating(rotationKey);
+            }
+            else if (wasRotating && !isRotating)
+            {
+                OnEndRotating(rotationKey);
+            }
+        }
+
+        private static void OnStartRotating(KeyBinding rotationKey)
+        {
+            if (numRotating == 0)
+            {
```

Comment:
Why is this (and the function down below) using a == 0 check (if it's only called internally in this class and is gated on transitions - i.e. false -> true, true->false?). It seems like this should only transition from 0 -> 1 and then 1 -> 0

Reply:
This is a refcount to make sure that cursor visibility and locking is handled correctly when both head and hands are rotating at the same time.
* Cursor should not be unlocked when one rotation stops while the other is still going.
* Cursor must be unlocked when all rotations have stopped.

It's still not 100% reliable because it only handles the input sim subsystem - if any other system uses these static cursor features it can mess up the cursor state too. We don't have a global API for managing this. But at least this common case of head + hands rotations works.